### ISSUE TITLE
Running terraform apply from plan out.

### ISSUE THREAD

**Issue by dsiperek-vendavo** [flaupretre/terraform-ssh-tunnel#19]
When running terraform apply from a plan out this module with not work.  Do you have any work arounds?

**Comment by flaupretre** [comment]
I agree with @seanamos' comment. I have no solution for this at the moment. A future solution may come from changes in the way terraform handles data sources or, maybe, from the transformation of this module into a provider. Anyway, it is not planned yet.

**Comment by dm3ch** [comment]
@flaupretre By the way, I have been thinking about the implementation as a provider for quite some time (haven't previous experience in writing provider) and still have no clue if it would really allow to bypass this problem.

The problem I see that we need to return local port number to use during connection from the provider. If I right understand the provider don't returns any data. So in this case usage of some data_source or something like this would be required. Wouldn't that create completely same issue, but instead of impossibility of tunnel start it would stack with port number inconsistency?

Maybe you have some thoughts, how it can be done or bypassed?

**Comment by Ezzahhh** [comment]
I suppose this must also mean that using this module with Atlantis is not possible? Given that Atlantis must output a tfplan and applies that tfplan - I don't believe it supports directly applying

**Comment by dm3ch** [comment]
> I suppose this must also mean that using this module with Atlantis is not possible? Given that Atlantis must output a tfplan and applies that tfplan - I don't believe it supports directly applying

In recent version of TF ephemeral resources was released. As far as I understand if tunnel would be created using ephemeral resource - it should work fine with Atlantis.

Here's an example of [provider that allows to create AWS SSM tunnels using ephemeral resources](https://github.com/dfns/terraform-provider-tunnel/pull/2). Technically it's possible to implement any other tunnel the same way.

**Comment by Ezzahhh** [comment]
@dm3ch Thanks for sharing that project. From my current testing it seems to work but I don't see the tunnel properly closed after terraform exits (Session Manager still reports that the tunnels are connected) - is there a way to ensure that ephemeral resources are appropriately closed on exit?

**Comment by dm3ch** [comment]
Hmm.

I haven't used that partifcular provider I mentioned on previous message but had played a little bit with https://github.com/ComplyCo/terraform-provider-aws-ssm-tunnels/pull/89 a while. 

That provider lacks Close logic at all. And while I was using it I haven't realised that.

But I believe this can be easily tested. But to test it right it's needed to understand you concerns. Are you concerned about local port not being closed or lack of ssm session termination on AWS side?

**Comment by Ezzahhh** [comment]
Yes, it's mainly just that the session is not closed on AWS side (it appears that the provider terminates the PID locally so the tunnel itself is closed) but when the PID is killed AWS seems not to realise this so it waits until the idle session termination kicks in (default 20min). It's mostly just a security concern but not necessarily a deal breaker as long as there aren't any session limits on AWS (which I don't think there is).

**Comment by arnaud-dfns** [comment]
Hi! We've added support for SSH tunnels in a recent PR in the [`dfns/tunnel`](https://github.com/dfns/terraform-provider-tunnel) provider. Both with `data` and `ephemeral` resources. https://github.com/dfns/terraform-provider-tunnel/pull/3
We will release a new version soon alongside the SSM session Termination bugfix that @Ezzahhh mentioned and other improvements.

**Comment by arnaud-dfns** [comment]
It's now available in release `v1.2.0` of `dfns/tunnel` provider! https://github.com/dfns/terraform-provider-tunnel/releases/tag/v1.2.0
